pkg/kv/kvtest/store.go

```diff
@@ -527,3 +529,155 @@ func testDeleteWhileIterPrefixSingleSequence(t *testing.T, ms MakeStore, sequenc
 		}
 	}
 }
+
+func testDeleteWhileIterSamePrefix(t *testing.T, ms MakeStore) {
+	// building the following kv keys sequence
+
+	//       pref			prefin			prefinin
+	//         |               |               |               |
+	// prefix: |-----------------------------------------------|
+	// internal-prefix:        |-------------------------------|
+	// internal-internal-prefix:               |---------------|
+
+	// Using the following keys (for some ints:L,M,N, and unique TESTID), that will provide access
+
+	// TESTID-pref-key-0000
+	// TESTID-pref-key-0001
+	// ...
+	// TESTID-pref-key-000<N>
+	// TESTID-prefin-key-0000
+	// TESTID-prefin-key-0001
+	// ...
+	// TESTID-prefin-key-000<M>
+	// TESTID-prefinin-key-0000
+	// TESTID-prefinin-key-0001
+	// ...
+	// TESTID-prefinin-key-000<L>
+
+	extPref := uniqueKey("pref")
+	inPref := uniqueKey("prefin")
+	ininPref := uniqueKey("prefinin")
+
+	allPrefs := [][]byte{extPref, inPref, ininPref}
+
+	// delete a subsection of the read prefix
+	testDeleteWhileIterSamePrefixSingleRun(t, ms, allPrefs, inPref, ininPref)
+
+	// scan a subsection of the delete prefix
+	testDeleteWhileIterSamePrefixSingleRun(t, ms, allPrefs, ininPref, inPref)
+
+	// scan and delete the same prefix
+	testDeleteWhileIterSamePrefixSingleRun(t, ms, allPrefs, ininPref, ininPref)
+}
+
+func testDeleteWhileIterSamePrefixSingleRun(t *testing.T, ms MakeStore, prefsToCreate [][]byte, readPref, delPref []byte) {
+	ctx := context.Background()
+	store := ms(t, ctx)
+	defer store.Close()
+
+	// Emptying the KV store from previous entries. This is essential in these tests as
+	// the verification relies on exact keys to exist
+	cleanIter, err := store.Scan(ctx, delPref)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer cleanIter.Close()
+
+	for cleanIter.Next() {
+		e := cleanIter.Entry()
+		if cleanIter != nil {
+			err = store.Delete(ctx, e.Key)
+			if err != nil {
+				t.Fatal(err)
+			}
+		}
+	}
```

Comment:
Interesting, you delete while you scan as a setup to a test where you check that. Maybe be on the safe side and aggregate the scan results in an array and delete afterwards?

Reply:
Agreed. Done